In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
df = pd.read_csv(r'C:\Users\ajkap\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [3]:
df = pd.read_csv(r'C:\Users\ajkap\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [4]:
path = r'C:\Users\ajkap\Instacart Basket Analysis'

In [5]:
os.path.join(path, '02 Data', 'Original Data', 'products.csv')

'C:\\Users\\ajkap\\Instacart Basket Analysis\\02 Data\\Original Data\\products.csv'

In [6]:
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [7]:
df_ords = pd.read_csv(os.path.join(path,'02 Data','Original Data','orders.csv'), index_col = False)

In [9]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [10]:
# create a dataframe
df_test = pd.DataFrame()

In [11]:
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
df_prods.dropna(subset = ['product_name'], inplace = True)

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_prods_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 Data Consistency Checks

# Question 2 part 2. The min and max of the days_since_prior_order column caught my attention. The fact the max is only 3 days since and the 50% amount is 7 seemed interesting to me. I would have though the 50% amount would be more in the 4-5 range. I also thought the 50% range being higher than the max which is 3.000 was unusual as well. The order_number column also has the max as being less than the 25% and 50% range which I did not think made sense. 

# Question 3. Check for mixed-type data in df_ords

In [29]:
# Mixed type data for df_ords
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [32]:
# Mixed type data for df_ords
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

# Question 4. I did not find any mixed-data types in the df_ords dataframe. If I did however, I would use the function df_test['mix'] = df_test['mix'].astype('str') to unmix the data and I could just replace the astype 'str' to int64 if I needed to switch the type to a numeric one.

# Question 5. Run a check for missing values

In [35]:
# Checking for missing values for df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [38]:
df_ords.head(30)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   
11   2168274        2    prior             1          2                 11   
12   1501582        2    prior             2          5                 10   
13   1901567        2    prior             3          1                 10   
14    738281        2    prior             4          2                 10   
15   1673511        2    prior             5          3                 11   
16   1199898        2    prior             6          2                  9   
17   3194192        2    prior             7          2                 12   
18    788338        2    prior             8          1                 15   
19   1718559        2    prior             9          2                  9   
20   1447487        2    prior            10          1                 11   
21   1402090        2    prior            11          1                 10   
22   3186735        2    prior            12          1                  9   
23   3268552        2    prior            13          4                 11   
24    839880        2    prior            14          3                 10   
25   1492625        2    train            15          1                 11   
26   1374495        3    prior             1          1                 14   
27    444309        3    prior             2          3                 19   
28   3002854        3    prior             3          3                 16   
29   2037211        3    prior             4          2                 18   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                     NaN  
27                     9.0  
28                    21.0  
29                    20.0

Question 5 Response. There are 206209 missing values in the days_since_prior_order column. It looks like NaN is with the order_number column equaling 1.It looks like the total number of user_id is 206209 which is the same as the number of missing values. This could suggest that each user_id that equals 1 has a missing value in the days_since_prior_order column. This makes sense since this was their first order and it is impossible to have something in those customers' days_since_prior_order column if they have only ever had 1 order to begin with. Therefore, it is accounted for why this column for those customers was left blank.

# Question 6 Address Missing Values


In [39]:
# Address Missing Values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]

In [40]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Question 6 Explanation. The missing value is consistent with customers making first time orders. The fact that these values are missing provides us information on customer orders specifically if the customer has ever placed an order with us or not. Because of this helpful information, I have decided not to remove or change these missing values in the dataframe. For further exploration it may be a good idea to notate that order_number columns that equal 1 represent new customers. 

# Question 7. Run a check for duplicate values 

In [44]:
#Duplicate check for df_ords data frame
df_ords_dups = df_ords[df_ords.duplicated()]

In [47]:
df_ords_dups.size

0

In [48]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Question 7 Explanation. I did not find any duplicates in the df_ords dataframe. 

#Question 8. Address the duplicates in an appropriate method. Since there were no duplicates found in the df_ords dataframe, I decided not to make any changes. If I had found duplicates I could use the code df.drop_duplicates() to get rid of any duplicates found. But before that I would use the code df_prods_clean to check how many rows and columns I had before getting rid of any duplicates. Then I can create a new dataframe without the dupcliates by using the code df_ords_clean_no_dups = df_ords_clean.drop_duplicates(). 

In [8]:
#Crosstabs
crosstab = pd.crosstab(ords_prods_merge['days_since_prior_order'], ords_prods_merge['order_number'], dropna = False)

NameError: name 'ords_prods_merge' is not defined

# Question 9. Exporting New Dataframes 

In [49]:
# Exporting df_ords as orders.checked 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [51]:
# Exporting df_prods as Products.checked
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))